plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt

```diff
@@ -0,0 +1,173 @@
+                    INTRODUCE_CONSTANT
+                )
+            }
+            else -> {
+                val options = ExtractionOptions(extractAsProperty = true)
+                val extractionData = ExtractionData(file, adjustedElements.toRange(), target, null, options)
+                ExtractionEngine(helper).run(editor, extractionData) {
+                    val property = it.declaration as KtProperty
+                    val descriptor = it.config.descriptor
+
+                    editor.caretModel.moveToOffset(property.textOffset)
+                    editor.selectionModel.removeSelection()
+                    if (editor.settings.isVariableInplaceRenameEnabled && !ApplicationManager.getApplication().isUnitTestMode) {
+                        with(PsiDocumentManager.getInstance(project)) {
+                            commitDocument(editor.document)
+                            doPostponedOperationsAndUnblockDocument(editor.document)
+                        }
+
+                        val introducer = KotlinInplacePropertyIntroducer(
+                            property = property,
+                            editor = editor,
+                            project = project,
+                            title = INTRODUCE_CONSTANT,
+                            doNotChangeVar = false,
+                            exprType = descriptor.returnType,
+                            extractionResult = it,
+                            availableTargets = listOf(ExtractionTarget.PROPERTY_WITH_GETTER)
+                        )
+                        introducer.performInplaceRefactoring(LinkedHashSet(getNameSuggestions(property) + descriptor.suggestedNames))
+                    } else {
+                        processDuplicatesSilently(it.duplicateReplacers, project)
+                    }
+                }
+            }
+        }
+    }
+
+    private fun validateElements(elements: List<PsiElement>): String? {
+        val errorMessage = validateExpressionElements(elements)
+        return when {
+            errorMessage != null -> errorMessage
+            elements.any { it.isNotConst() } -> KotlinBundle.message("error.text.can.t.introduce.constant.for.this.expression.because.not.constant")
+            else -> null
+        }
+    }
+
+    private fun PsiElement.isNotConst() = !isConst()
+
+    private fun PsiElement.isConst(): Boolean {
+        if (this is KtBinaryExpression) {
+            var isConst = true
+            object : KotlinRecursiveElementVisitor() {
+                override fun visitKtElement(element: KtElement) {
+                    super.visitKtElement(element)
+                    if (element !is KtOperationReferenceExpression && element !is KtBinaryExpression) {
+                        isConst = isConst && element.isConst()
+                    }
+                }
+            }.visitBinaryExpression(this)
+
+            return isConst
+        } else {
+            return this is KtConstantExpression
```

Comment:
not sure about whitespaces / new lines

Reply:
I put each type on a seperate line because otherwise the line would be very long. I could do 
```kotlin
        } else {
            return this is KtConstantExpression || this is KtLiteralStringTemplateEntry
                   || (this is KtStringTemplateExpression && !this.hasInterpolation())
        }
```

Reply:
I would put the 3 distinct logical conditions on separate lines.

Alternatively (not sure if this is in line with the IJ codebase) I prefer to be explicit when doing complex Boolean expressions:
```kt
return when (this) {
    is KtConstantExpression -> true
    is KtLiteralStringTemplateEntry -> true
    is KtStringTemplateExpression -> !this.hasInterpolation()
    else -> false
}
```
This has the added benefit of having a nice place to put comments on each case explaining them. And the code after the arrows have place to grow.